multivariate_normal_model.cpp

```diff
@@ -1,106 +1,52 @@
 #include "multivariate_normal_model.h"
 #include "opencv2/imgproc.hpp"
 
-MultivariateNormalModel::MultivariateNormalModel()
-    : is_trained_{false}
-{}
 
-bool MultivariateNormalModel::isTrained() const
-{ return is_trained_; }
+MultivariateNormalModel::MultivariateNormalModel(const cv::Mat& samples)
+{
+  performTraining(samples);
+}
+
 
 void MultivariateNormalModel::performTraining(const cv::Mat& samples)
 {
-  cv::Mat covar;
-  cv::calcCovarMatrix(samples, covar, mu_, cv::COVAR_NORMAL | cv::COVAR_COLS);
-  cv::invert(covar, inv_covar_, cv::DECOMP_SVD);
-
-  samples_ = samples.clone();
-  is_trained_ = true;
+  cv::calcCovarMatrix(samples, covariance_, mean_, cv::COVAR_NORMAL | cv::COVAR_COLS);
+  cv::invert(covariance_, inverse_covariance_, cv::DECOMP_SVD);
 }
 
+
 cv::Mat MultivariateNormalModel::computeMahalanobisDistances(const cv::Mat& image) const
 {
-  cv::Mat image_64;
-  image.convertTo(image_64, CV_64FC3);
+  constexpr double dist_to_uint8_scale = 1000.0;
 
-  cv::Mat mahalanobis_dist(image.size(), CV_64FC1);
+  cv::Mat mahalanobis_img(image.size(), CV_8UC1);
 
-  image_64.forEach<cv::Vec3d>(
-      [&](cv::Vec3d& pixel, const int pos[])
+  using Pixel = cv::Vec<uint8_t, 3>;
+  image.forEach<Pixel>(
```

Comment:
Hmm, hmm.. Jeg lurer på om jeg gjorde noe profiling på `cv::Mat::forEach` en gang (kanskje da jeg startet på CUDA studies for multimodal?), og da virket det på meg som at i hvert fall min datamaskin valgte å ikke parallellisere noe særlig.. Jeg husker ikke hvor store bilder det var. Så det er jo kult å lære bort at funksjonen finnes, men på den annen side er den jo også ganske rar. Så et innspill fra meg til en annen gang er å kanskje velge å gå for en tradisjonell, mer intuitiv, dobbel forløkke.